### DISCUSSION TITLE
Fast Multiple Thumbnails Per Image

### DISCUSSION THREAD

**Discussion by awesomecoder111** [libvips/libvips#2094] [Q&A]
I'm generating 5 thumbnails for a jpg image. The original image dimensions can vary widely in size from 300 to 15000 pixels (300 <= width, height <= 15000). The thumbnails will be around 1200x1000, 600x400, 250x250 etc.

Option 1:
* Compute shrink as [here](https://github.com/libvips/libvips/blob/df3b80fa40f4800e6949d356c800891af81537db/libvips/resample/thumbnail.c#L471): 
* Load image using vips_image_new_from_file with shrink [here](https://github.com/libvips/libvips/blob/df3b80fa40f4800e6949d356c800891af81537db/libvips/resample/thumbnail.c#L1040)
* for each desired thumbnail: vips_thumbnail_image and then write it out to disk

Option 2:
* call vips_thumbnail 5 times in parallel for each thumbnail.

Questions:
1. Am I using the right operations to make option 1 as fast as it can be? Option 2 could be faster but is resource heavy.
2. My concern with Option 1 is that I see  [documentation](https://libvips.github.io/libvips/API/current/libvips-resample.html#vips-thumbnail-image) mentions vips_thumbnail_image can "produce poorer quality output". Will this be an issue.
3. Should "sequential" access be enabled for the image loading in this scenario for option 1?

This is similar to https://github.com/libvips/php-vips/issues/103, but I don't see these options mentioned there.

**Comment by jcupitt** [comment]
Hi @awesomecoder111,

Because of the way libvips works, multiple parallel vipsthumbnail is by far the fastest. It shouldn't be too resource-heavy. At the CLI, for example, you can do:

```
$ vipsheader wtc.jpg
wtc.jpg: 9372x9372 uchar, 3 bands, srgb, jpegload
$ time parallel vipsthumbnail wtc.jpg -s {} -o wtc-{}.jpg ::: 1200 600 300 150 75
real	0m0.540s
user	0m1.838s
sys	0m0.152s
```

Five thumbnails in 500ms. `vips_thumbnail_image()` is very slow and only for emergencies. You need to use `vips_thumbnail()` to get all the nice shrink-on-load stuff.

**Comment by awesomecoder111** [comment]
Thanks @jcupitt!

- `vips_thumbnail_image() is very slow and only for emergencies` doesn't option 1 handle this by doing the shrink on load manually?
- say I didn't do option 2 in parallel, do you suspect calling vips_thumbnail 5 times sequentially would be better than option 1?

**Comment by jcupitt** [comment]
Ah is this only for jpg? True, your option 1 would be equivalent then, and it's not much code to copy-paste.

Poorer quality output is for things like PDF and SVG where `vips_thumbnail` will be able to compute the load resolution and not need to rely on downsampling.

Yes, `sequential` should help a bit.

Still, I would just run `thumbnnail` five times in parallel. If you are extremely resource constrained (under 100mb of ram for your whole system) then just run it five times sequentially.

The best thing of course would be to do some experiments. You should be able to mock up and test a range of solutions very quickly in something like Python.

**Comment by awesomecoder111** [comment]
Thanks @jcupitt I'm finding option 1 to be faster than option 2
For a 5000x3000 image:
- option 1: 0.75user 0.17system 0.28real 324%CPU 217maxresident MB
- option 2: 0.81user 0.38system 0.46real 259%CPU 207maxresident MB

For a 13000x10000 image:
- option 1: 2.25user 0.41system 1.05real 252%CPU 1003maxresident MB
- option 2: 2.55user 1.61system 1.23real 338%CPU 924maxresident MB

I'm running experiments using pyvips on a 2019 macbook pro with 8 cores, and generating 4 thumbnails. For option 2, I'm using gnu parallel to call my python script. I'm surprised since you suggested that option 2 should be faster?

For your reference when i run your command `time parallel vipsthumbnail wtc.jpg -s {} -o wtc-{}.jpg ::: 1200 600 300 150` on my 5000x3000 jpg image here are the times `0.53user 0.27system 0.35real`, and running `vipsheader` on it gives `3 bands, srgb, jpegload`. (Note that the thumbnail sizes for me are a bit different than 1200, 600 etc. and for one of them I'm doing a crop).

**Reply by jcupitt** [reply]
It's a while since I tested it, perhaps things have shifted. If you share a benchmark program I could time it as well.
